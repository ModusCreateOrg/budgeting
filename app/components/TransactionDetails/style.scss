@import 'theme/variables';

.transaction__details {
  text-align: left;
}

.neg {
  color: $red;
}

.pos {
  color: $green;
}
