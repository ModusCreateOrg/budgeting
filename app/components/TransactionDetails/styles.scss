.title{
  display: flex;
  flex-direction: column;
  align-items: center;
}

h1 h2 {
  flex: 1;
  text-align: center;
}

.negative  {
  color: red;
}

.positive {
  color: green;
}

.button {
  flex: 1;
  display: flex;
  justify-content: center;
  font-size: 25px;
  padding: 3px;
  background-color: red;
  border: solid 2px #c2185b;
  border-radius: 5px;
  color: #fff;
  width: 120px;
}
