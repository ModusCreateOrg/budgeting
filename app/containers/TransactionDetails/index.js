// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { scaleOrdinal, schemeCategory20 } from 'd3';
import { injectAsyncReducers } from 'store';
import formatAmountPercentageContribution from 'utils/formatAmountPercentageContribution';
import { shuffle } from 'utils/array';
import { getTransactionById, getInflowBalance, getOutflowBalance } from 'selectors/transactions';
import { getCategories } from 'selectors/categories';
import type { Transaction } from 'modules/transactions';
import transactionReducer from 'modules/transactions';
import categoryReducer from 'modules/categories';
import TransactionDetail from 'components/TransactionDetail';
import PieChart from 'components/PieChart';

const outflowScheme = shuffle([...schemeCategory20.slice(0, 4), ...schemeCategory20.slice(5)]);
const inflowScheme = ['#bcbd22', '#bcbd22', '#2ca02c']; // inflow always green

type TransactionDetailsProps = {
  transaction: Transaction,
  balance: number,
  categories: Object,
};

injectAsyncReducers({
  transactions: transactionReducer,
  categories: categoryReducer,
});

class TransactionDetails extends React.Component<TransactionDetailsProps> {
  static defaultProps = {
    transaction: {},
    balance: 0,
    categories: {},
  };

  getColorScheme = (transaction: Transaction) => {
    const key = transaction.value < 0 ? 'outflow' : 'inflow';
    return this.color[key].domain([0, 1]);
  };

  color = {
    inflow: scaleOrdinal(inflowScheme),
    outflow: scaleOrdinal(outflowScheme),
  };

  render() {
    const { transaction, balance, categories } = this.props;
    const remainingBalance = balance - Math.abs(transaction.value);
    const remainingPercentage = formatAmountPercentageContribution(remainingBalance, balance, 2, false).text;
    const data = [
      {
        description: `${transaction.description}: `,
        value: Math.abs(transaction.value),
        id: transaction.id,
      },
      {
        description: `${transaction.value < 0 ? 'Remainig Outflow' : 'Remaining Inflow'} (${remainingPercentage}): `,
        value: remainingBalance,
        id: transaction.id + 1,
      },
    ];

    return (
      <section>
        <TransactionDetail transaction={transaction} balance={balance} categories={categories} />
        <PieChart data={data} dataLabel="description" dataKey="id" color={this.getColorScheme(transaction)} />;
      </section>
    );
  }
}

const mapStateToProps = (state, { match }) => {
  const transactionId = parseInt(match.params.id, 10);
  const transaction = getTransactionById(state, transactionId);
  const balance = Math.abs(transaction.value > 0 ? getInflowBalance(state) : getOutflowBalance(state));
  return {
    transaction,
    balance,
    categories: getCategories(state),
  };
};

export default connect(mapStateToProps)(TransactionDetails);
