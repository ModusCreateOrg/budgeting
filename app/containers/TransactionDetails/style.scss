@import 'theme/variables';

.transactionDetails {
  display: flex;
  flex-flow: column;
  flex: 1;
  border: 1px solid $gray;
  border-collapse: collapse;
  font-size: 0.8em;
  padding-left: 10px;

  button {
    margin-top: 10px;
    font-size: 1.2em;
    width: 10%;
  }

  h1 {
    margin-bottom: 0px;
  }

  .neg {
    color: $red;
  }
  .pos {
    color: $green;
  }

  @media only screen and (max-width: $screen-small) {
    button {
      margin-top: 10px;
      font-size: 1em;
      width: 20%;
    }
  }
}
