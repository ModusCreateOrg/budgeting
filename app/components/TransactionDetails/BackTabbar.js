// @flow
import * as React from 'react';
import styles from './style.scss';
import { withRouter } from 'react-router'

const BackTabbar = (props, context) => (
  <div className={styles.tabbar}>
    <button 
      onClick={(e) => {props.history.goBack();}}
      className={`${styles.navLink}`}>
        Back
    </button>
  </div>
);

const BackTabbarWithRouter = withRouter(BackTabbar)

export default BackTabbarWithRouter;
