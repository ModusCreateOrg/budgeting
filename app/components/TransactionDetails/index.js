import * as React from 'react';
import DonutChart from 'components/DonutChart';
import styles from './styles.scss';

type TransactionDetailsProps = {
  isNegative: boolean;
  percent: number;
  transaction: Transaction;
  data: Transaction[];
  history: object;
};

const TransactionDetails = ({isNegative, percent, transaction, data, history }) => (
  <section>
    <div className={styles.title}>
      <h1 className={isNegative ? styles.negative : styles.positive} >
        {transaction.description}
      </h1>
      <h2 className={isNegative ? styles.negative : styles.positive}>
        {isNegative ? '-' : ''}{percent + ' %'}
      </h2>
      <a className={styles.button} onClick={history.goBack}>Go back</a>
      <DonutChart data={data} dataLabel="description" dataKey="id"/>
    </div>
  </section>
);

export default TransactionDetails;
