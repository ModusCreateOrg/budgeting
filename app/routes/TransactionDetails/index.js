import React, { Component } from 'react';
import Chunk from 'components/Chunk';

const loadTransactionDetailsContainer = () =>
  import('containers/TransactionDetails' /* webpackChunkName: "transactiondetails" */);

class TransactionDetails extends Component<{}> {
  render() {
    return <Chunk load={loadTransactionDetailsContainer} match={this.props.match} />;
  }
}

export default TransactionDetails;
