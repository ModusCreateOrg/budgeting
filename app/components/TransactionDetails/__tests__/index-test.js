import React from 'react';
import render from 'react-test-renderer';
import TransactionDetails from '../';

it('renders correctly', () => {
  const mockTransaction = {
    id: 1,
    description: "Trader Joe's food",
    value: -423.34,
    categoryId: 1,
  };
  const mockPercentage = '30';

  const tree = render.create(<TransactionDetails transaction={mockTransaction} percentage={mockPercentage} />).toJSON();
  expect(tree).toMatchSnapshot();
});
