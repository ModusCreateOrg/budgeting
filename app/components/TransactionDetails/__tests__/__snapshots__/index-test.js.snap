// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section>
  <div
    className="title"
  >
    <h1
      className="negative"
    >
      Trader Joe's food
    </h1>
    <h2
      className="negative"
    >
      -
      9.23 %
    </h2>
    <a
      className="button"
      onClick={undefined}
    >
      Go back
    </a>
    <div
      className="donutChart"
    >
      <svg
        className="mainSvg"
        height={316}
        viewBox="-8 -8 316 316"
        width={316}
      >
        <g
          transform="translate(150,150)"
        >
          <path
            d="M9.184850993605149e-15,-150A150,150,0,0,1,82.17194942404177,-125.49012203298201L41.085974712020885,-62.74506101649101A75,75,0,0,0,4.592425496802574e-15,-75Z"
            fill="#bcbd22"
          />
          <path
            d="M82.17194942404177,-125.49012203298201A150,150,0,1,1,-1.6078131593583422e-13,-150L-8.039065796791711e-14,-75A75,75,0,1,0,41.085974712020885,-62.74506101649101Z"
            fill="#ff7f0e"
          />
        </g>
      </svg>
      <ul
        className="legend"
      >
        <li
          style={
            Object {
              "color": "#bcbd22",
            }
          }
        >
          <span>
            Trader Joe's food
          </span>
          <span
            className="value"
          >
             
            $423.34
             
          </span>
        </li>
        <li
          style={
            Object {
              "color": "#ff7f0e",
            }
          }
        >
          <span>
            Other
          </span>
          <span
            className="value"
          >
             
            $4,164.73
             
          </span>
        </li>
      </ul>
    </div>
  </div>
</section>
`;
